### PR TITLE
vulkan/context: propagate --gpu-sw when using vulkan

### PR DESCRIPTION
**Pull request by llyyr**
Broken since day 1

### CODE REVIEW

**Comment by kasper93** [comment]
Could you add it also to `pl_d3d11_create`?

File: video/out/vulkan/context.c [hunk 1]
```diff
@@ -31,6 +31,7 @@ struct vulkan_opts {
     int queue_count;
     bool async_transfer;
     bool async_compute;
+    bool allow_sw;
```
Comment (kasper93) [inline video/out/vulkan/context.c:34]: Could we use whole name `allow_software` and put it in alphabetical order?

File: video/out/vulkan/context.c [hunk 2]
```diff
@@ -238,6 +239,7 @@ pl_vulkan mppl_create_vulkan(struct vulkan_opts *opts,
         .num_opt_extensions = MP_ARRAY_SIZE(opt_extensions),
         .features = &features,
         .device_name = is_uuid ? NULL : opts->device,
+        .allow_software = opts->allow_sw,
```
Comment (kasper93) [inline video/out/vulkan/context.c:242]: Put it in the same place as in the `pl_vulkan_params`, meaning before `async_transfer`

**Comment by llyyr** [comment]
> Could you add it also to `pl_d3d11_create`?

I added it but this is actually a change in behavior. libplacebo sets `allow_software = true` by default for d3d11 contexts, but mpv defaults to it being disabled. I don't know if this matters though

**Comment by kasper93** [comment]
> > Could you add it also to `pl_d3d11_create`?
> 
> I added it but this is actually a change in behavior. libplacebo sets `allow_software = true` by default for d3d11 contexts, but mpv defaults to it being disabled. I don't know if this matters though

I don't know.  Software implementation is used over RDP, but in practice, it is the same story as using software impl in any other context. It will be slower than normal.

**Comment by llyyr** [comment]
Does it need to be mentioned in interface-changes?

**Comment by kasper93** [comment]
> Does it need to be mentioned in interface-changes?

In theory not, but maybe it will help affected users discover this? I know some use mpv over RDP, there were reports related to it. In the same time, it was already documented.

Maybe `vo_gpu` should also be fixed.

File: video/filter/vf_gpu_vulkan.c [hunk 3]
```diff
@@ -79,7 +79,7 @@ static struct offscreen_ctx *vk_offscreen_ctx_create(struct mpv_global *global,
         goto error;
 
     struct vulkan_opts *vk_opts = mp_get_config_group(NULL, global, &vulkan_conf);
-    vk->vulkan = mppl_create_vulkan(vk_opts, vk->vkinst, vk->pllog, VK_NULL_HANDLE);
+    vk->vulkan = mppl_create_vulkan(ractx->swapchain, vk->vkinst, vk->pllog, VK_NULL_HANDLE);
```
Comment (kasper93) [inline video/filter/vf_gpu_vulkan.c:82]: We likely don't have swapchain here. Have you tested it? Also above line is dead code now.

Comment (llyyr) [reply video/filter/vf_gpu_vulkan.c]: I can't test vf_gpu with vulkan atm so I don't wanna make changes to it blindly, for now I just added allow_software to mppl_create_vulkan

**Comment by llyyr** [comment]
https://mpv.io/manual/master/#options-d3d11-warp

I think it makes sense for gpu-sw to only affect opengl/vulkan contexts, as for d3d11 it's controlled by d3d11-warp, but I don't think any options from `struct d3d11_device_opts` are mapped to libplacebo, so this is better done in another PR by someone who can actually test it

**Comment by kasper93** [comment]
So, should `gpu-sw` be reverted to its original name `opengl-sw` and new flag `vulkan-sw` be added if this should be handle on per api basis?

**Comment by llyyr** [comment]
> So, should `gpu-sw` be reverted to its original name `opengl-sw` and new flag `vulkan-sw` be added if this should be handle on per api basis?

I don't think it's necessary, the only reason d3d11-warp exists is because somebody thought there was some value in forcing warp vs just allowing it. And you can force it a specific device with vulkan-device anyway

**Review by kasper93 — APPROVED** [review]

**Pull request merged**